From bb6d9d0efb845cf2027dbe6ecba085dcfcfdd561 Mon Sep 17 00:00:00 2001
From: Andy Ross <andrew.j.ross@intel.com>
Date: Tue, 12 Aug 2014 12:09:26 -0700
Subject: [PATCH 089/117] atomisp_gmin_platform: Correct TI PMIC implementation

Schematics for actual devices show two different LDOs (10 and 11)
wired to the sensor 1.8v inputs.  Drive those instead of LDO10 alone.
Clean up usage a little bit to clarify register contents.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
---
 .../x86/platform/intel-mid/atomisp_gmin_platform.c | 25 ++++++++++++----------
 1 file changed, 14 insertions(+), 11 deletions(-)

diff --git a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
index dd3565b..db4921d 100644
--- a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
+++ b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
@@ -50,12 +50,14 @@ EXPORT_SYMBOL(spid);
 
 /* TI SND9039 PMIC register hackery */
 #define LDO9_REG	0x49
-#define LDO9_2P8V_ON	0x2f
-#define LDO9_2P8V_OFF	0x2e
-
 #define LDO10_REG	0x4a
-#define LDO10_1P8V_ON	0x59
-#define LDO10_1P8V_OFF	0x58
+#define LDO11_REG	0x4b
+
+#define LDO_2P8V_ON	0x2f /* 0x2e selects 2.85V ...      */
+#define LDO_2P8V_OFF	0x2e /* ... bottom bit is "enabled" */
+
+#define LDO_1P8V_ON	0x59 /* 0x58 selects 1.80V ...      */
+#define LDO_1P8V_OFF	0x58 /* ... bottom bit is "enabled" */
 
 struct gmin_subdev {
 	struct v4l2_subdev *subdev;
@@ -434,10 +436,11 @@ int gmin_v1p8_ctrl(struct v4l2_subdev *subdev, int on)
 	}
 
 	if (pmic_id == PMIC_TI) {
-		if (on)
-			return intel_soc_pmic_writeb(LDO10_REG, LDO10_1P8V_ON);
-		else
-			return intel_soc_pmic_writeb(LDO10_REG, LDO10_1P8V_OFF);
+		int val = on ? LDO_1P8V_ON : LDO_1P8V_OFF;
+		int ret = intel_soc_pmic_writeb(LDO10_REG, val);
+		if (!ret)
+			ret = intel_soc_pmic_writeb(LDO11_REG, val);
+		return ret;
 	}
 
 	return -EINVAL;
@@ -487,9 +490,9 @@ int gmin_v2p8_ctrl(struct v4l2_subdev *subdev, int on)
 
 	if (pmic_id == PMIC_TI) {
 		if (on)
-			return intel_soc_pmic_writeb(LDO9_REG, LDO9_2P8V_OFF);
+			return intel_soc_pmic_writeb(LDO9_REG, LDO_2P8V_ON);
 		else
-			return intel_soc_pmic_writeb(LDO9_REG, LDO9_2P8V_OFF);
+			return intel_soc_pmic_writeb(LDO9_REG, LDO_2P8V_OFF);
 	}
 
 	return -EINVAL;
-- 
1.9.3

